### ISSUE TITLE
[feature request] Log-determinant for symmetric positive definite matrices

### ISSUE THREAD

**Issue by calincru** [pytorch/pytorch#22848]
## 🚀 Feature

As of now, one cannot leverage the positive-definitiveness of a matrix when computing log-determinants. The current `torch.logdet` works for general square matrices and thus uses LU in forward and SVD in backward.  

For SPD matrices it can be implemented more efficiently via a Cholesky decomposition in forward and its inverse in backward.  The latter can be computed very efficiently via `torch.cholesky_inverse` (note that its batch variant is not implemented yet: https://github.com/pytorch/pytorch/issues/7500).

CC @vishwakftw

**Comment by calincru** [comment]
For the moment, one can do the following:

```python
import torch
from torch.autograd import Function

class PLogDet(Function):
  @staticmethod
  def forward(ctx, x):
    l = torch.cholesky(x)    
    ctx.save_for_backward(l)
    return 2 * l.diagonal(dim1=-2, dim2=-1).log().sum(-1)

  @staticmethod
  def backward(ctx, g):
    l, = ctx.saved_tensors
    n = l.shape[-1]
    # use cholesky_inverse once pytorch/pytorch/issues/7500 is solved
    return torch.cholesky_solve(torch.eye(n, out=l.new(n, n)), l)

plogdet = PLogDet.apply
```
(I wrote it directly here so it might contain typos.)

**Comment by vishwakftw** [comment]
Hi @calincru, thanks for opening the feature request. We can probably add an argument (`sym_pos_def`  for example) to `logdet`, and add an implementation.

**Comment by fmassa** [comment]
@vishwakftw I have a question about adding an argument to `logdet`.

Should it be an extra argument or should we add a whole new function for this?
We currently have some functions (like `eig` and `symeig`) which I believe do the same thing, but assume different structures for the input data, so it might be worth considering if we should standardize somehow or not

**Comment by vishwakftw** [comment]
@fmassa an implementation based on Cholesky could be provided not just for logdet, but also for det and slogdet. Adding a separate function for symmetric positive definite matrices would create 3 extra functions, and too many names to remember and then use.

Havint said that, even I’m not sure as to how to proceed. TensorFlow defines logdet for only SPD matrices (while we don’t), and slogdet is used for general matrices.

**Comment by Gsunshine** [comment]
@fmassa An extra argument may be better. We only need to remember one name and the hint could tell us how to use. Very easy! 

@vishwakftw The implementation from TensorFlow could be a good reference. There are at least two reasons to follow it, 
① No Free Lunch : ). Usually we gain benefits from specialization. logdet will be faster than original version.
② It is more easier to remember the api for users who switch from TensorFlow to Pytorch.

**Comment by ezyang** [comment]
We discussed this in the triage meeting. The request seems reasonable but we're not sure what to do. If we had a way of O(1) checking if a matrix was SPD we could just add a conditional to do the right thing in this situation... but we don't, and adding labeling for properties of matrices would be a pretty substantial project, much more than what you originally asked for.

The keyword argument idea is not unreasonable, but it's a slippery slope: there are lots of operations where you might also have a more efficient implementation if a matrix has some special properties. Maybe it's OK for this case, and the patch will look nice and it'll be the only boolean flag we end up adding to logdet. We're just not sure if this is the case or not :)

**Comment by calincru** [comment]
@ezyang  Why not add a separate function then?

**Comment by ezyang** [comment]
Well, as mentioned earlier, it seems that there's a family of three
functions that we'd have to add separate functions for.  Keyword
argument might be locally right, but it's hard to say if it's globally
right without doing some more analysis over all the linear algebra
functions we have.

Excerpts from Călin Cruceru's message of 2019-07-22 11:33:24 -0700:
> @ezyang  Why not add a separate function then?
>

**Comment by calincru** [comment]
@ezyang Sorry, I don't get it -- what's the problem with having three different functions?

I personally disagree with the "slippery slope" thing.  IMHO, you should let the demand drive new specialized implementations: no one asks for `logdet` of SPD matrices, don't add it; more and more people request it, add it. It's the reason we have `eig` and `symeig`, as @fmassa pointed out, but there's no `spdeig`, despite there being algorithms that are faster or more stable for positive definite matrices (see, e.g., https://epubs.siam.org/doi/pdf/10.1137/0613074, Section 4.3).

This means that I totally understand if you postpone this for now due to lack of a significant number of requests.

**Comment by ezyang** [comment]
Ah, well, maybe there is not a problem with three more functions :)  It
is mostly a matter of discoverability.

Re demand, I agree with you!  But another thing that we like to do is, even if
demand is low, if someone cares enough about a feature to implement it,
we like to come with a determination about whether or not we would
invest the time to review and merge it.  I think in this case the
answer is yes--we just need to agree on the API :)

Excerpts from Călin Cruceru's message of 2019-07-25 09:06:31 -0700:
> @ezyang Sorry, I don't get it -- what's the problem with having three different functions?
> 
> I personally disagree with the "slippery slope" thing.  IMHO, you should let the demand drive new specialized implementations: no one asks for `logdet` of SPD matrices, don't add it; more and more people request it, add it. It's the reason we have `eig` and `symeig`, as @fmassa pointed out, but there's no `spdeig`, despite there being algorithms that are faster or more stable for positive definite matrices (see, e.g., https://epubs.siam.org/doi/pdf/10.1137/0613074, Section 4.3).
> 
> This means that I totally understand if you postpone this for now due to lack of a significant number of requests.
>

**Comment by Gsunshine** [comment]
Personally, I still agree with adding an kwarg, only for `logdet`. People ask for some features unless it have to be. It is not 'a slippery slope'. I could get the point to keep APIs clear and well-conducted. But ..., it is usually evolving with demands.

> Ah, well, maybe there is not a problem with three more functions :) It is mostly a matter of discoverability. Re demand, I agree with you! But another thing that we like to do is, even if demand is low, if someone cares enough about a feature to implement it, we like to come with a determination about whether or not we would invest the time to review and merge it. I think in this case the answer is yes--we just need to agree on the API :) Excerpts from Călin Cruceru's message of 2019-07-25 09:06:31 -0700:
> […](#)
> @ezyang Sorry, I don't get it -- what's the problem with having three different functions? I personally disagree with the "slippery slope" thing. IMHO, you should let the demand drive new specialized implementations: no one asks for `logdet` of SPD matrices, don't add it; more and more people request it, add it. It's the reason we have `eig` and `symeig`, as @fmassa pointed out, but there's no `spdeig`, despite there being algorithms that are faster or more stable for positive definite matrices (see, e.g., https://epubs.siam.org/doi/pdf/10.1137/0613074, Section 4.3). This means that I totally understand if you postpone this for now due to lack of a significant number of requests.

**Comment by ezyang** [comment]
I'll stick my neck out here, and say we will accept a PR that adds `sym_pos_def: bool` argument to `logdet` which uses an alternate implementation.

**Comment by vishwakftw** [comment]
I’m working on adding batching for det/logdet/slogdet. I can work on this once that is done.

**Comment by Gsunshine** [comment]
I report this bug #23473 . It might be helpful for the discussion.

**Comment by Balandat** [comment]
I think having separate functions is fine. I think what we really should have are improved docs that group operators together for better discoverability. 

E.g. of the sort

# Linear algebra functions

## By operation

### Eigenvalues and Eigenvectors

#### General matrices
- torch.eig
#### Symmetric matrices
- torch.symeig
#### Positive Semidefinite matrices
...

### Determinants
...

## By matrix type

### General matrices

#### Eigenvalues and Eigenvectors
- torch.eig
- torch.symeig

#### Determinants
...
### Symmetric matrices
...
### Positive Semidefinite matrices

**Comment by gchanan** [comment]
I think we should discuss the more general approach to this issue in https://github.com/pytorch/pytorch/issues/28341.

For the plan of record on this one, let's match `symeig` and have a new function.

**Comment by Lezcano** [comment]
I think we won't be pursuing this for now. Now, this would be a nice example of what we could do with StructuredTensors, as mentioned above (https://github.com/pytorch/pytorch/issues/28341). Also, once we have the LDL decomposition (https://github.com/pytorch/pytorch/pull/69828), this could be extended to symmetric matrices.

In the meantime, the code proposed above
```python
2 * torch.linalg.cholesky(A).diagonal(dim1=-2, dim2=-1).log().sum(-1)
```
is differentiable and works with batches, so it may be used as a replacement for `torch.linalg.slogdet` when the matrix is SPD.

As a side-note, `linalg.lu` does not use SVD in the backward anymore, but `lu_solve`.
